File: web/src/generator/node.ts
import { Expression, LiteralExpression, NameExpression, Prop } from "./expression";

export class Node {
	readonly start: number;
	constructor(start: number) {
		this.start = start;
	}
}

export class GrammarDeclaration extends Node {
	readonly rules: readonly RuleDeclaration[];
	readonly topRules: readonly RuleDeclaration[];
	readonly tokens: TokenDeclaration | null;
	readonly localTokens: readonly LocalTokenDeclaration[];
	readonly context: ContextDeclaration | null;
	readonly externalTokens: readonly ExternalTokenDeclaration[];
	readonly externalSpecializers: readonly ExternalSpecializeDeclaration[];
	readonly externalPropSources: readonly ExternalPropSourceDeclaration[];
	readonly precedences: PrecDeclaration | null;
	readonly mainSkip: Expression | null;
	readonly scopedSkip: readonly {
		expr: Expression;
		topRules: readonly RuleDeclaration[];
		rules: readonly RuleDeclaration[];
	}[];
	readonly dialects: readonly Identifier[];
	readonly externalProps: readonly ExternalPropDeclaration[];
	readonly autoDelim: boolean;

	constructor(
		start: number,
		rules: readonly RuleDeclaration[],
		topRules: readonly RuleDeclaration[],
		tokens: TokenDeclaration | null,
		localTokens: readonly LocalTokenDeclaration[],
		context: ContextDeclaration | null,
		externalTokens: readonly ExternalTokenDeclaration[],
		externalSpecializers: readonly ExternalSpecializeDeclaration[],
		externalPropSources: readonly ExternalPropSourceDeclaration[],
		precedences: PrecDeclaration | null,
		mainSkip: Expression | null,
		scopedSkip: readonly {
			expr: Expression;
			topRules: readonly RuleDeclaration[];
			rules: readonly RuleDeclaration[];
		}[],
		dialects: readonly Identifier[],
		externalProps: readonly ExternalPropDeclaration[],
		autoDelim: boolean,
	) {
		super(start);
		this.rules = rules;
		this.topRules = topRules;
		this.tokens = tokens;
		this.localTokens = localTokens;
		this.context = context;
		this.externalTokens = externalTokens;
		this.externalSpecializers = externalSpecializers;
		this.externalPropSources = externalPropSources;
		this.precedences = precedences;
		this.mainSkip = mainSkip;
		this.scopedSkip = scopedSkip;
		this.dialects = dialects;
		this.externalProps = externalProps;
		this.autoDelim = autoDelim;
	}
	toString(): string {
		return Object.values(this.rules).join("\n");
	}
}

export class RuleDeclaration extends Node {
	readonly id: Identifier;
	readonly props: readonly Prop[];
	readonly params: readonly Identifier[];
	readonly expr: Expression;

	constructor(
		start: number,
		id: Identifier,
		props: readonly Prop[],
		params: readonly Identifier[],
		expr: Expression,
	) {
		super(start);
		this.id = id;
		this.props = props;
		this.params = params;
		this.expr = expr;
	}
	toString(): string {
		return (
			this.id.name + (this.params.length ? `<${this.params.join()}>` : "") + " -> " + this.expr
		);
	}
}

export class PrecDeclaration extends Node {
	readonly items: readonly { id: Identifier; type: "left" | "right" | "cut" | null }[];

	constructor(
		start: number,
		items: readonly { id: Identifier; type: "left" | "right" | "cut" | null }[],
	) {
		super(start);
		this.items = items;
	}
}

export class TokenPrecDeclaration extends Node {
	readonly items: readonly (NameExpression | LiteralExpression)[];

	constructor(start: number, items: readonly (NameExpression | LiteralExpression)[]) {
		super(start);
		this.items = items;
	}
}

export class TokenConflictDeclaration extends Node {
	readonly a: NameExpression | LiteralExpression;
	readonly b: NameExpression | LiteralExpression;

	constructor(
		start: number,
		a: NameExpression | LiteralExpression,
		b: NameExpression | LiteralExpression,
	) {
		super(start);
		this.a = a;
		this.b = b;
	}
}

export class TokenDeclaration extends Node {
	readonly precedences: readonly TokenPrecDeclaration[];
	readonly conflicts: readonly TokenConflictDeclaration[];
	readonly rules: readonly RuleDeclaration[];
	readonly literals: readonly LiteralDeclaration[];

	constructor(
		start: number,
		precedences: readonly TokenPrecDeclaration[],
		conflicts: readonly TokenConflictDeclaration[],
		rules: readonly RuleDeclaration[],
		literals: readonly LiteralDeclaration[],
	) {
		super(start);
		this.precedences = precedences;
		this.conflicts = conflicts;
		this.rules = rules;
		this.literals = literals;
	}
}

export class LocalTokenDeclaration extends Node {
	readonly precedences: readonly TokenPrecDeclaration[];
	readonly rules: readonly RuleDeclaration[];
	readonly fallback: { readonly id: Identifier; readonly props: readonly Prop[] } | null;

	constructor(
		start: number,
		precedences: readonly TokenPrecDeclaration[],
		rules: readonly RuleDeclaration[],
		fallback: { readonly id: Identifier; readonly props: readonly Prop[] } | null,
	) {
		super(start);
		this.precedences = precedences;
		this.rules = rules;
		this.fallback = fallback;
	}
}

export class LiteralDeclaration extends Node {
	readonly literal: string;
	readonly props: readonly Prop[];

	constructor(start: number, literal: string, props: readonly Prop[]) {
		super(start);
		this.literal = literal;
		this.props = props;
	}
}

export class ContextDeclaration extends Node {
	readonly id: Identifier;
	readonly source: string;

	constructor(start: number, id: Identifier, source: string) {
		super(start);
		this.id = id;
		this.source = source;
	}
}

export class ExternalTokenDeclaration extends Node {
	readonly id: Identifier;
	readonly source: string;
	readonly tokens: readonly { id: Identifier; props: readonly Prop[] }[];
	readonly conflicts: readonly Identifier[];

	constructor(
		start: number,
		id: Identifier,
		source: string,
		tokens: readonly { id: Identifier; props: readonly Prop[] }[],
		conflicts: readonly Identifier[],
	) {
		super(start);
		this.id = id;
		this.source = source;
		this.tokens = tokens;
		this.conflicts = conflicts;
	}
}

export class ExternalSpecializeDeclaration extends Node {
	readonly type: "extend" | "specialize";
	readonly token: Expression;
	readonly id: Identifier;
	readonly source: string;
	readonly tokens: readonly { id: Identifier; props: readonly Prop[] }[];

	constructor(
		start: number,
		type: "extend" | "specialize",
		token: Expression,
		id: Identifier,
		source: string,
		tokens: readonly { id: Identifier; props: readonly Prop[] }[],
	) {
		super(start);
		this.type = type;
		this.token = token;
		this.id = id;
		this.source = source;
		this.tokens = tokens;
	}
}

export class ExternalPropSourceDeclaration extends Node {
	readonly id: Identifier;
	readonly source: string;

	constructor(start: number, id: Identifier, source: string) {
		super(start);
		this.id = id;
		this.source = source;
	}
}

export class ExternalPropDeclaration extends Node {
	readonly id: Identifier;
	readonly externalID: Identifier;
	readonly source: string;

	constructor(start: number, id: Identifier, externalID: Identifier, source: string) {
		super(start);
		this.id = id;
		this.externalID = externalID;
		this.source = source;
	}
}

export class Identifier extends Node {
	readonly name: string;

	constructor(start: number, name: string) {
		super(start);
		this.name = name;
	}
	toString(): string {
		return this.name;
	}
}

File: dotnet/Rezel/Generator/Node.cs
namespace Rezel.Generator;

public class Node
{
    public readonly int Start;

    public Node(int start)
    {
        Start = start;
    }
}

public sealed class ScopedSkipEntry
{
    public readonly Expression Expr;
    public readonly RuleDeclaration[] TopRules;
    public readonly RuleDeclaration[] Rules;

    public ScopedSkipEntry(Expression expr, RuleDeclaration[] topRules, RuleDeclaration[] rules)
    {
        Expr = expr;
        TopRules = topRules;
        Rules = rules;
    }
}

public sealed class PrecItem
{
    public readonly Identifier Id;
    public readonly string? Type;

    public PrecItem(Identifier id, string? type)
    {
        Id = id;
        Type = type;
    }
}

public sealed class TokenEntry
{
    public readonly Identifier Id;
    public readonly Prop[] Props;

    public TokenEntry(Identifier id, Prop[] props)
    {
        Id = id;
        Props = props;
    }
}

public sealed class GrammarDeclaration : Node
{
    public readonly RuleDeclaration[] Rules;
    public readonly RuleDeclaration[] TopRules;
    public readonly TokenDeclaration? Tokens;
    public readonly LocalTokenDeclaration[] LocalTokens;
    public readonly ContextDeclaration? Context;
    public readonly ExternalTokenDeclaration[] ExternalTokens;
    public readonly ExternalSpecializeDeclaration[] ExternalSpecializers;
    public readonly ExternalPropSourceDeclaration[] ExternalPropSources;
    public readonly PrecDeclaration? Precedences;
    public readonly Expression? MainSkip;
    public readonly ScopedSkipEntry[] ScopedSkip;
    public readonly Identifier[] Dialects;
    public readonly ExternalPropDeclaration[] ExternalProps;
    public readonly bool AutoDelim;

    public GrammarDeclaration(
        int start,
        RuleDeclaration[] rules,
        RuleDeclaration[] topRules,
        TokenDeclaration? tokens,
        LocalTokenDeclaration[] localTokens,
        ContextDeclaration? context,
        ExternalTokenDeclaration[] externalTokens,
        ExternalSpecializeDeclaration[] externalSpecializers,
        ExternalPropSourceDeclaration[] externalPropSources,
        PrecDeclaration? precedences,
        Expression? mainSkip,
        ScopedSkipEntry[] scopedSkip,
        Identifier[] dialects,
        ExternalPropDeclaration[] externalProps,
        bool autoDelim
    ) : base(start)
    {
        Rules = rules;
        TopRules = topRules;
        Tokens = tokens;
        LocalTokens = localTokens;
        Context = context;
        ExternalTokens = externalTokens;
        ExternalSpecializers = externalSpecializers;
        ExternalPropSources = externalPropSources;
        Precedences = precedences;
        MainSkip = mainSkip;
        ScopedSkip = scopedSkip;
        Dialects = dialects;
        ExternalProps = externalProps;
        AutoDelim = autoDelim;
    }
}

public sealed class RuleDeclaration : Node
{
    public readonly Identifier Id;
    public readonly Prop[] Props;
    public readonly Identifier[] Params;
    public readonly Expression Expr;

    public RuleDeclaration(
        int start,
        Identifier id,
        Prop[] props,
        Identifier[] parameters,
        Expression expr
    ) : base(start)
    {
        Id = id;
        Props = props;
        Params = parameters;
        Expr = expr;
    }
}

public sealed class PrecDeclaration : Node
{
    public readonly PrecItem[] Items;

    public PrecDeclaration(int start, PrecItem[] items) : base(start)
    {
        Items = items;
    }
}

public sealed class TokenPrecDeclaration : Node
{
    public readonly Expression[] Items;

    public TokenPrecDeclaration(int start, Expression[] items) : base(start)
    {
        Items = items;
    }
}

public sealed class TokenConflictDeclaration : Node
{
    public readonly Expression A;
    public readonly Expression B;

    public TokenConflictDeclaration(int start, Expression a, Expression b) : base(start)
    {
        A = a;
        B = b;
    }
}

public sealed class TokenDeclaration : Node
{
    public readonly TokenPrecDeclaration[] Precedences;
    public readonly TokenConflictDeclaration[] Conflicts;
    public readonly RuleDeclaration[] Rules;
    public readonly LiteralDeclaration[] Literals;

    public TokenDeclaration(
        int start,
        TokenPrecDeclaration[] precedences,
        TokenConflictDeclaration[] conflicts,
        RuleDeclaration[] rules,
        LiteralDeclaration[] literals
    ) : base(start)
    {
        Precedences = precedences;
        Conflicts = conflicts;
        Rules = rules;
        Literals = literals;
    }
}

public sealed class LocalTokenDeclaration : Node
{
    public readonly TokenPrecDeclaration[] Precedences;
    public readonly RuleDeclaration[] Rules;
    public readonly TokenEntry? Fallback;

    public LocalTokenDeclaration(
        int start,
        TokenPrecDeclaration[] precedences,
        RuleDeclaration[] rules,
        TokenEntry? fallback
    ) : base(start)
    {
        Precedences = precedences;
        Rules = rules;
        Fallback = fallback;
    }
}

public sealed class LiteralDeclaration : Node
{
    public readonly string Literal;
    public readonly Prop[] Props;

    public LiteralDeclaration(int start, string literal, Prop[] props) : base(start)
    {
        Literal = literal;
        Props = props;
    }
}

public sealed class ContextDeclaration : Node
{
    public readonly Identifier Id;
    public readonly string Source;

    public ContextDeclaration(int start, Identifier id, string source) : base(start)
    {
        Id = id;
        Source = source;
    }
}

public sealed class ExternalTokenDeclaration : Node
{
    public readonly Identifier Id;
    public readonly string Source;
    public readonly TokenEntry[] Tokens;
    public readonly Identifier[] Conflicts;

    public ExternalTokenDeclaration(
        int start,
        Identifier id,
        string source,
        TokenEntry[] tokens,
        Identifier[] conflicts
    ) : base(start)
    {
        Id = id;
        Source = source;
        Tokens = tokens;
        Conflicts = conflicts;
    }
}

public sealed class ExternalSpecializeDeclaration : Node
{
    public readonly string Type;
    public readonly Expression Token;
    public readonly Identifier Id;
    public readonly string Source;
    public readonly TokenEntry[] Tokens;

    public ExternalSpecializeDeclaration(
        int start,
        string type,
        Expression token,
        Identifier id,
        string source,
        TokenEntry[] tokens
    ) : base(start)
    {
        Type = type;
        Token = token;
        Id = id;
        Source = source;
        Tokens = tokens;
    }
}

public sealed class ExternalPropSourceDeclaration : Node
{
    public readonly Identifier Id;
    public readonly string Source;

    public ExternalPropSourceDeclaration(int start, Identifier id, string source) : base(start)
    {
        Id = id;
        Source = source;
    }
}

public sealed class ExternalPropDeclaration : Node
{
    public readonly Identifier Id;
    public readonly Identifier ExternalID;
    public readonly string Source;

    public ExternalPropDeclaration(int start, Identifier id, Identifier externalID, string source) : base(start)
    {
        Id = id;
        ExternalID = externalID;
        Source = source;
    }
}

public sealed class Identifier : Node
{
    public readonly string Name;

    public Identifier(int start, string name) : base(start)
    {
        Name = name;
    }
}
